File: Cpp/Rehashing.cpp
#include <vector>

/**
 * Definition of ListNode
 */
class ListNode {
public:
    int val;
    ListNode *next;
    ListNode(int val) {
        this->val = val;
        this->next = NULL;
    }
};

class Solution {
public:
    /**
     * @param hashTable: A list of The first node of linked list
     * @return: A list of The first node of linked list which have twice size
     */
    vector<ListNode*> rehashing(vector<ListNode*> hashTable) {
        // write your code here
        int len = hashTable.size(), newlen = 2 * len;
        vector<ListNode*> vec(newlen);
        for (int i = 0; i < len ; i++) {
            ListNode *cnt = hashTable[i];
            while (cnt != NULL) {
                int key = (cnt->val % newlen + newlen) % newlen;
                //ListNode *res = vec[key];
                ListNode *new_node = new ListNode(cnt->val);
                if (vec[key] == NULL) {
                    vec[key] = new_node;
                } else {
                    ListNode *res = vec[key];
                    while (res->next != NULL) {
                        res = res->next;
                    }
                    res->next = new_node;
                }
                cnt = cnt->next;
            }
        }
        return vec;
    }
};

File: Java/Rehashing.java
/**
 * Definition for ListNode
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param hashTable: A list of The first node of linked list
     * @return: A list of The first node of linked list which have twice size
     */    
    public ListNode[] rehashing(ListNode[] hashTable) {
        // write your code here
        int capacity = hashTable.length * 2;
        ListNode[] res = new ListNode[capacity];
        for (int i = 0; i < hashTable.length; i ++) {
            ListNode cur = hashTable[i];
            while (cur != null) {
                int key = (cur.val % capacity + capacity) % capacity;
                ListNode new_node = new ListNode(cur.val);
                if (res[key] == null) {
                    res[key] = new_node;
                } else {
                    ListNode temp = res[key];
                    while(temp.next != null) {
                        temp = temp.next;
                    }
                    temp.next = new_node;
                }
                cur = cur.next;
            }
        }
        return res;
    }
};

